sw/device/lib/testing/mock_mmio.h

```diff
@@ -81,131 +192,286 @@ class MmioTest {
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_MASK8_AT(dev, offset, ...) \
+  EXPECT_MASK_INTERAL_(8, dev, offset, __VA_ARGS__)
+
+/**
+ * Expect an unspecified 16-bit read to the device |dev| at the given offset,
+ * followed by a write to the same location, with the same value but with some
+ * bits changed; the remaining bits must be untouched.
+ *
+ * The changed bits are specified by a |std::initializer_list<MaskedBitField>|.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_MASK16_AT(dev, offset, ...) \
+  EXPECT_MASK_INTERAL_(16, dev, offset, __VA_ARGS__)
+
+/**
+ * Expect an unspecified 32-bit read to the device |dev| at the given offset,
+ * followed by a write to the same location, with the same value but with some
+ * bits changed; the remaining bits must be untouched.
+ *
+ * The changed bits are specified by a |std::initializer_list<MaskedBitField>|.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_MASK32_AT(dev, offset, ...) \
+  EXPECT_MASK_INTERAL_(32, dev, offset, __VA_ARGS__)
+
+/**
+ * Expect an unspecified 8-bit read at the given offset, followed by a write to
+ * the same location, with the same value but with some bits changed; the
+ * remaining bits must be untouched.
+ *
+ * The changed bits are specified by a |std::initializer_list<MaskedBitField>|.
+ *
+ * This function is only available in tests using a fixture that derives
+ * |MmioTest|.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_MASK8(offset, ...) \
+  EXPECT_MASK32_AT(this->dev(), offset, __VA_ARGS__)
+
+/**
+ * Expect an unspecified 16-bit read at the given offset, followed by a write to
+ * the same location, with the same value but with some bits changed; the
+ * remaining bits must be untouched.
+ *
+ * The changed bits are specified by a |std::initializer_list<MaskedBitField>|.
+ *
+ * This function is only available in tests using a fixture that derives
+ * |MmioTest|.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_MASK16(offset, ...) \
+  EXPECT_MASK32_AT(this->dev(), offset, __VA_ARGS__)
+
+/**
+ * Expect an unspecified 32-bit read at the given offset, followed by a write to
+ * the same location, with the same value but with some bits changed; the
+ * remaining bits must be untouched.
+ *
+ * The changed bits are specified by a |std::initializer_list<MaskedBitField>|.
+ *
  * This function is only available in tests using a fixture that derives
- * |DeviceTest|.
+ * |MmioTest|.
  *
  * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
  * calls.
  */
-#define EXPECT_WRITE32(offset, value) \
-  EXPECT_WRITE32_AT(this->dev(), offset, value);
+#define EXPECT_MASK32(offset, ...) \
```

Comment:
To be honest I think there it could be very useful to have a function with the same semantics as this mask in `mmio.h`, the only thing I am slightly weary about is, whether it will be more costly in terms of performance/space than our current shifting method:

```
  // Set baudrate, enable RX and TX, configure parity                                                                                                                                          
  uint32_t reg = (nco_masked << UART_CTRL_NCO_OFFSET);
  reg |= (1u << UART_CTRL_TX);
  reg |= (1u << UART_CTRL_RX);
  if (config->parity_enable == kDifUartEnable) {
    reg |= (1u << UART_CTRL_PARITY_EN);
  }
  if (config->parity == kDifUartParityOdd) {
    reg |= (1u << UART_CTRL_PARITY_ODD);
  }
  mmio_region_write32(uart->base_addr, UART_CTRL_REG_OFFSET, reg);
```

Reply:
I think MASK32 is mostly suited for expecting simple changes in bitfields, which I think are a bit more common. Something like a control register has a different feel to it altogether, because you're going to do a batch write of a bunch of fields and aren't doing read-modify-write.

Perhaps EXPECT_MASK should be called EXPECT_RMW...

Reply:
Also, the code above is a bit dangerous. It is fine in init routine, where you don't care about the current register state, but otherwise you have to have semantics of the `EXPECT_MASK32` anyway, otherwise you can end up annihilating the rest of the register.